plugins/UsersManager/Model.php

```diff
@@ -245,11 +245,14 @@ public function userEmailExists($userEmail)
     public function addUserAccess($userLogin, $access, $idSites)
     {
         foreach ($idSites as $idsite) {
-            $this->getDb()->insert(Common::prefixTable("access"),
-                array("idsite" => $idsite,
-                      "login"  => $userLogin,
-                      "access" => $access)
-            );
+            $count = $this->getDb()->fetchOne("SELECT count(*) FROM " . Common::prefixTable("site"). " WHERE idsite = ?",$idsite);
+            if($count != 0){
+                 $this->getDb()->insert(Common::prefixTable("access"),
+                    array("idsite" => $idsite,
+                         "login"  => $userLogin,
+                         "access" => $access)
+                );
+            }
```

Comment:
Our models are supposed to be very stupid, not having any logic. It's not really like a model more like a DAO. Also the users manager should not know about the sites manager or the data structure of sites (in the case it should not having to know that there is a `site` table etc).

Ideally the logic would be implemented in the `UsersManager\API` method `setUserAccess()`. There you could do something like 

```
foreach ($idSites as $idsite) {
    $site = Piwik\API\Request::processRequest('SitesManager.getSiteFromId', array('idSite' => $idSite));
    if (!empty($site)) {
         $this->model->addUserAccess($userLogin, $access, $idSite);
   }
}
```

I have not tested the code but something like that should work. I know it's a bit annoying as one would have to do it also around `deleteUserAccess()` etc. Maybe the code could be simplified eg `$idSites = $this->getOnlyActuallyExistingIdSites($idSites)` at the beginning of `setUserAccess` API method.